code/datums/mergers/_merger.dm

```diff
@@ -83,56 +83,78 @@
+	var/list/found_turfs = list()
 	if(origin)
-		tips[origin] = NORTH|EAST|SOUTH|WEST
-		new_members[origin] = NONE
-	while(length(tips))
-		var/atom/focus = tips[length(tips)]
-		var/dirs_to_check = tips[focus]
-		tips.len--
+		var/turf/starting = get_turf(origin)
+		check_turf(starting, found_turfs, NONE)
+	for(var/i = 1; i <= length(found_turfs), i++)
+		var/turf/focus = found_turfs[i]
+		var/list/focus_packet = found_turfs[focus]
+		var/dirs_checked = focus_packet[MERGE_TURF_PACKET_DIR]
 		for(var/dir in GLOB.cardinals)
-			if(!(dirs_to_check & dir))
+			if(dirs_checked & dir)
 				continue
 			var/turf/location = get_step(focus, dir)
-			if(!location || checked_turfs[location])
+			if(!location)
 				continue
-			checked_turfs[location] = TRUE
-			for(var/i in location)
-				var/atom/movable/thing = i
-				if(!merged_typecache[thing.type])
-					continue
-				if(attempt_merge_proc && !call(thing, attempt_merge_proc)(src, new_members))
-					continue
-				if(thing.mergers && thing.mergers[id] != src)
-					var/datum/merger/existing = thing.mergers[id]
-					qdel(src)
-					existing.Refresh()
+			if(!check_turf(location, found_turfs, dir))
+				if(QDELETED(src))
 					return
-				new_members[focus] |= dir // This is not a list, value of the members list is a bitfield of dirs
-				var/next_dirs = turn(dir, 180)
-				new_members[thing] |= next_dirs
-				next_dirs = ~next_dirs
-				var/existing = tips[thing] || (NORTH|EAST|SOUTH|WEST)
-				tips[thing] = existing & next_dirs
-
-	var/list/leaving_members = members - new_members
+				continue
+			focus_packet[MERGE_TURF_PACKET_DIR] |= dir
+
+	// Now that we have an idea of our connecting directions, build the fresh members list
+	var/list/fresh_members = list()
+	for(var/turf/location as anything in found_turfs)
+		var/list/turf_packet = found_turfs[location]
+		var/connected_dirs = turf_packet[MERGE_TURF_PACKET_DIR]
+		for(var/datum/member as anything in turf_packet[MERGE_TURF_PACKET_ATOMS])
+			fresh_members[member] = connected_dirs
+
+	var/list/leaving_members = members - fresh_members
 	for(var/atom/thing as anything in leaving_members)
 		RemoveMember(thing)
 
-	var/list/joining_members = new_members - members
+	var/list/joining_members = fresh_members - members
 	for(var/atom/thing as anything in joining_members)
 		AddMember(thing, joining_members[thing])
 
 	// They may not need a full update but the connected dirs could change
-	for(var/atom/thing as anything in new_members)
-		members[thing] = new_members[thing]
+	for(var/atom/thing as anything in fresh_members)
+		members[thing] = fresh_members[thing]
 
 	SEND_SIGNAL(src, COMSIG_MERGER_REFRESH_COMPLETE, leaving_members, joining_members)
 
 	if(!length(members))
 		qdel(src)
 
+// Checks to see if the passed in location contains something interesting to us. If it does, return TRUE, otherwise return false
+// If it is interesting, we add it to our processing list
+/datum/merger/proc/check_turf(turf/location, list/found_turfs, asking_from)
```

Comment:
While in most other cases I would agree with this seperation into a proc, this particular system has high sensitivity to lag, enough that proc overhead could become meaningful as well as the additional list operation required by this change in method. Could you profile the difference between these two versions of Refresh()? The only change the original should require is some extra handling of the origin to be equivalent in function.

Reply:
Sure, I can do that. The two aren't completely functionally the same, but I see your concern. Gimme a sec

Reply:
Here's the cost of refreshing 10k times on a group with 306 members 
![image](https://user-images.githubusercontent.com/58055496/151727952-36bf5985-2c06-4853-9b36-216b1253455f.png)

This is without any modifications to refresh. I can't easily retain functionality between the two, because the old system of sharing was at its base kinda borked, but I can fix the double stacked origin issue.
That profile looks like this 
![image](https://user-images.githubusercontent.com/58055496/151728696-105a4ed1-de5a-4f2f-840d-3319a25809e0.png)

Here's a profile with the same code, but against 439 merger items double stacked on their tiles
![image](https://user-images.githubusercontent.com/58055496/151729303-54216737-1a89-41b8-82b4-1800bafd2f47.png)

Reply:
Seems fine